### PR TITLE
Add a test per issue https://github.com/halide/Halide/issues/2731 .

### PR DESCRIPTION
**Pull request by zvookin**
Also serves as documnetation on passing parameters to JITted routines
and avoiding recompilation for each realize.

Suggestions on further comments for documentation are particularly
welcome.

### CODE REVIEW

File: test/performance/thread_safe_jit.cpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+}
+
+void same_func_per_thread() {
+    std::thread threads[16];
+    test_func test;
+
+    {
+        std::lock_guard<std::mutex> lock(compiler_mutex);
```
Comment (dsharletg) [inline test/performance/thread_safe_jit.cpp]: This is unnecessary in this case I think. Harmless, but slightly misleading too.

File: test/performance/thread_safe_jit.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+}
+
+void separate_func_per_thread() {
+    std::thread threads[16];
+
+    for (auto &thread : threads) {
+        thread = std::thread(separate_func_per_thread_executor,
+                             (int)(&thread - threads));
```
Comment (dsharletg) [inline test/performance/thread_safe_jit.cpp]: This is a bit clever for something so simple... seems like a regular for (int i = ...) loop that uses threads[i] would be simpler.

(same below.)

File: test/performance/thread_safe_jit.cpp (hunk 1 above)
Comment (zvookin) [reply test/performance/thread_safe_jit.cpp]: Added a comment to indicate why it is included even though in some sense it is not strictly necessary.

File: test/performance/thread_safe_jit.cpp (hunk 2 above)
Comment (zvookin) [reply test/performance/thread_safe_jit.cpp]: I have a few options:
1) `for (int i = 0; i < 16; i++) {` which repeats the "16" and thus seems suboptimal.
2) `for (int i = 0; i < (int)(sizeof(threads)/sizeof(threads[0])); i++) {` which is idiomatic but ugly.
3) The above which I felt was the least of the evils.

I could also do `#define THREADS 16`, but that also seems a bit dated.

Comment (steven-johnson) [reply test/performance/thread_safe_jit.cpp]: Eh, gotta side with Z on this one; this seems like a textbook case of where the range-based for loop is simpler.

**Pull request merged**